src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs

```diff
@@ -116,28 +180,52 @@ public CheckResult Check(string expressionText, RecordType parameterType = null,
         /// <param name="parameterType">types of additional args to pass.</param>
         /// <param name="options">parser options to use.</param>
         /// <returns></returns>
-        public CheckResult Check(ParseResult parse, RecordType parameterType = null, ParserOptions options = null)
+        public CheckResult Check(ParseResult parse, RecordType parameterType, ParserOptions options = null)
+        {
+            var bindingConfig = new BindingConfig(allowsSideEffects: options?.AllowsSideEffects == true);
+            var symbols = SymbolTable.NewFromRecord(parameterType);
+
+            return Check2(parse, bindingConfig, symbols);
+        }
+
+        public CheckResult Check(
+            string expressionText,
+            ParserOptions options = null,
+            ReadOnlySymbolTable symbolTable = null)
+        {
+            var parse = Parse(expressionText, options);
+            return Check(parse, options, symbolTable);
+        }
+
+        public CheckResult Check(
+            ParseResult parse,
+            ParserOptions options = null,
+            ReadOnlySymbolTable symbolTable = null)
         {
             var bindingConfig = new BindingConfig(allowsSideEffects: options?.AllowsSideEffects == true);
-            return CheckInternal(parse, parameterType, bindingConfig);
+
+            return Check2(parse, bindingConfig, symbolTable);
         }
 
-        private CheckResult CheckInternal(ParseResult parse, RecordType parameterType, BindingConfig bindingConfig)
+        private CheckResult Check2(
+            ParseResult parse,
+            BindingConfig bindingConfig = null,
+            ReadOnlySymbolTable symbolTable = null)
         {
-            parameterType ??= RecordType.Empty();
+            var startHash = symbolTable?.VersionHash;
 
             // Ok to continue with binding even if there are parse errors. 
             // We can still use that for intellisense. 
-
-            var resolver = CreateResolver();
+            var resolver = CreateResolverInternal(symbolTable);
+            
             var glue = CreateBinderGlue();
 
             var binding = TexlBinding.Run(
                 glue,
                 parse.Root,
                 resolver,
                 bindingConfig,
-                ruleScope: parameterType._type,
+                ruleScope: null,
```

Comment:
>null [](http://example.com/codeflow?start=27&length=4)

Is this intentional/correct?

Reply:
Yes. 
We've removed the parameterType altogether.
This information is now passed in via symbolTable, and via INameResolver

Reply:
it's an optional param with a null default, any reason not to just leave it out?

Reply:
I just wanted to be explicit - rule scope distinction was such a critical issue here.